keras_hub/src/models/yolo_v8/non_max_suppression.py

```diff
@@ -0,0 +1,502 @@
+        represents box coordinates in yxyx format.
+      scores: a tensor of rank 1 or higher with a shape of [..., num_boxes].
+      max_output_size: a scalar integer tensor representing the maximum number
+        of boxes to be selected by non max suppression.
+      iou_threshold: a float representing the threshold for deciding whether boxes
+        overlap too much with respect to IoU (intersection over union).
+      score_threshold: a float representing the threshold for box scores. Boxes
+        with a score that is not larger than this threshold will be suppressed.
+      tile_size: an integer representing the number of boxes in a tile, i.e.,
+        the maximum number of boxes per image that can be used to suppress other
+        boxes in parallel; larger tile_size means larger parallelism and
+        potentially more redundant work.
+
+    Returns:
+      selected_box_args: a tensor with a shape of [..., num_boxes] representing the
+        indices selected by non-max suppression. The leading dimensions
+        are the batch dimensions of the input boxes. All numbers are within
+        [0, num_boxes). For each image (i.e., selected_box_args[i]), only the first
+        num_valid[i] indices (i.e., selected_box_args[i][:num_valid[i]]) are valid.
+      num_valid: a tensor of rank 0 or higher with a shape of [...]
+        representing the number of valid indices in selected_box_args.
+        Its dimensions are the batch dimensions of the input boxes.
+    """  # noqa: E501
+    batch_dims = ops.shape(boxes)[:-2]
+    num_boxes = boxes.shape[-2]
+    boxes = ops.reshape(boxes, [-1, num_boxes, 4])
+    scores = ops.reshape(scores, [-1, num_boxes])
+    batch_size = boxes.shape[0]
+    if score_threshold != float("-inf"):
+        score_mask = ops.cast(scores > score_threshold, scores.dtype)
+        scores *= score_mask
+        box_mask = ops.expand_dims(ops.cast(score_mask, boxes.dtype), 2)
+        boxes *= box_mask
+
+    scores, boxes, sorted_indices = _sort_scores_and_boxes(scores, boxes)
+
+    pad = (
+        math.ceil(max(num_boxes, max_output_size) / tile_size) * tile_size
+        - num_boxes
+    )
+    boxes = ops.pad(ops.cast(boxes, "float32"), [[0, 0], [0, pad], [0, 0]])
+    scores = ops.pad(ops.cast(scores, "float32"), [[0, 0], [0, pad]])
+    num_boxes_after_padding = num_boxes + pad
+    num_iterations = num_boxes_after_padding // tile_size
+
+    def _loop_cond(unused_boxes, unused_threshold, output_size, tile_arg):
+        return ops.logical_and(
+            ops.min(output_size) < ops.cast(max_output_size, "int32"),
+            ops.cast(tile_arg, "int32") < num_iterations,
+        )
+
+    def suppression_loop_body(boxes, iou_threshold, output_size, tile_arg):
+        return _suppression_loop_body(
+            boxes, iou_threshold, output_size, tile_arg, tile_size
+        )
+
+    selected_boxes, _, output_size, _ = ops.while_loop(
```

Comment:
1. This works with all backends and I used the same one for retinanet but for now if we can include as this is detachable from model easily and not involved in trained and only while predictions.

2.  We have to check if its bottleneck compared to torch and then we can make some changes as they use loop based approach rather than `ops.while_loop`.
3. We have to consider normal `for loop` as conversion to onxx and then tensorrt and previous tensorflow models had problems especially for NonMaxSupression, so we have to keep in considerations to this use case and write it NonMaxSupression.

Reference: https://github.com/pytorch/vision/blob/acbfd8d94d10f989f4540252e92e8855c19f7ff7/torchvision/models/detection/retinanet.py#L518

Reply:
Hi @divyashreepathihalli from what I understand NMS seems to be programmed sequentially because of the necessity of carrying a state of unprocessed boxes. From what I have seen, NMS seems to be usually implemented using a while / for loop since boxes are greedily chosen and removed. I don't know if vectorizing this would decrease the computation time given the sequential nature of NMS. Moreover, [pytorch's NMS implementation](https://github.com/pytorch/vision/blob/main/torchvision/csrc/ops/cpu/nms_kernel.cpp) is done in c++ using a double for loop across boxes. Do let me know how you would like me to proceed.